Introduce parsing to metric units

neupatryk opened the pull request:
In this pr I am introducing script for parsing imperial units to metric based if enabled in preferences. Script is crawling through data files and search for the regex pattern where imperial unit is preceded by the number or set of numbers. 
When pattern is detected, it is wrapped by new tag, this is why changes in the data files are required. 

## What is still to be done if logic is to be approved
- Handle inches and weight
- Handle data different than bestiary
- Handle case "number-feet" and simillar

## Questions 
Currently in repo there is conversion from miles to kilometers where 1mile = 1.6km. In my monster manual conversion is 1mile = 1.5km. Which one should be correct?

revilowaldow commented:
It's more a style thing, and I'm sure Giddy will comment on the wider merits of this PR.

The system in use in the books is "US Customary" not "Imperial" which was an old British Empire system.
They're different so using `imperial` everywhere would be unideal.

See: 
https://en.wikipedia.org/wiki/Imperial_units
https://en.wikipedia.org/wiki/United_States_customary_units
https://en.wikipedia.org/wiki/Imperial_and_US_customary_measurement_systems

### Other comments:
I note you've not handled weights (lb. -> kg) at all, is this intentional (following [localized products](https://www.gf9.com/Default.aspx?tabid=348&art_id=5526) which don't change the weights), or were you only looking at distance first? 
Temperature as doesn't seem to have been considered, in e.g. `recipe`s, but also in e.g. DMG Extreme heat/cold hazard rules

You've also edited the attacks and senses for creatures, but not touched the speed data, which, like e.g. spell ranges, could be handled as a display only as they're implicitly in feet unless [otherwise declared](https://github.com/TheGiddyLimit/5etools-utils/blob/6509315c2d615783debcb934ab5cee732d6f196d/schema-template/spells/spells.json#L80C21-L80C21).

neupatryk commented:
> It's more a style thing, and I'm sure Giddy will comment on the wider merits of this PR.
> 
> The system in use in the books is "US Customary" not "Imperial" which was an old British Empire system. They're different so using `imperial` everywhere would be unideal.
> 
> See: https://en.wikipedia.org/wiki/Imperial_units https://en.wikipedia.org/wiki/United_States_customary_units https://en.wikipedia.org/wiki/Imperial_and_US_customary_measurement_systems
> 
> ### Other comments:
> I note you've not handled weights (lb. -> kg) at all, is this intentional (following [localized products](https://www.gf9.com/Default.aspx?tabid=348&art_id=5526) which don't change the weights), or were you only looking at distance first? Temperature also doesn't seem to have been considered, in e.g. `recipe`s, but also in e.g. DMG Extreme heat/cold hazard rules
> 
> You've also edited the attacks and senses for creatures, but not touched the speed data, which, like e.g. spell ranges, could be handled as a display only as they're implicitly in feet unless [otherwise declared](https://github.com/TheGiddyLimit/5etools-utils/blob/6509315c2d615783debcb934ab5cee732d6f196d/schema-template/spells/spells.json#L80C21-L80C21).

So going one by one:

Name "imperial" can be easly changed. My first idea was to use "unit", but then I saw that it is already used. Do you have some idea on better name for this tag? 

Missing data as written in the pr description is on purpose. I am not that familliar with your repository and I wanted to have initial confirm if my proposition is worth moving forward. If I would run scripts on all files it would make this pr really big. I have todos for that and currently only distances in only bestiary is parsed as example.

I am not sure about, what you mean by speed, but there is already code for changing speed I just plugged flag to use preference setting instead always false. So monster speed should work without changes as you store numbers not strings. I added screenshots of parsed aboleth in the description.

About declared distances, I was not aware about them and I would need to think what could be done, but as there is amount field and unit field I think it should be also possible to handle.

TheGiddyLimit commented:
fine enough as a concept, though plenty to be picking over implementation-wise

---

> In my monster manual, conversion is 1mile = 1.5km.

what version is that, for reference? Happy to adjust 5etools' numbers, they're mostly extrapolated from the one Reddit thread [noted as a comment in the source](https://old.reddit.com/r/dndnext/comments/6gkuec). Screenshots/pictures/page numbers/ISBN(!)/... much appreciated

---

not a big fan of `@imperial`; given `@unit` is taken, I reckon `@wm` ("weights and measures"). Since these are going to end up absolutely everywhere, the shorter the better

not keen on your current tag format, though. I'd like to see something more structured, e.g.:

```
{@imperial 60 ft.} -> {@wm 60|ft} -> rendered as "60 ft."
{@imperial 60 feet} -> {@wm 60|ft|p} // "p" = "plural" -> rendered as "60 feet"
{@imperial 1 foot} -> {@wm 1|ft|s} // "s" = "singular" -> rendered as "1 foot"
```

or 

```
{@imperial 60 ft.} -> {@wm 60|ft} -> rendered as "60 ft."
{@imperial 60 feet} -> {@wm 60|ft|v} // "v" = "verbose"; explicitly converted to plural as `60` -> rendered as "60 feet"
{@imperial 1 foot} -> {@wm 1|ft|v} // "v" = "verbose"; explicitly converted to singular as `1` -> rendered as "1 foot"
```

there should never be regex in the renderer; we expect to be rendering well-formed data, not parsing it

the above might have some holes, but the gist should be: splitting on `|` should yield:

- a single number
- a single unit specifier, picked from a minimal set (i.e. we should not allow both `ft` and `feet`)
- anything else we need (e.g. the rendering flags above; see `@area` for inspiration/prior work)

---

additional things which need considered:

- There was some recent chat about converting `monster.senses` to something more structured, which would eliminate the need for adding tags there. Would therefore prefer if those were _not_ tagged
- Homebrew builder
- Text converter

additional things which need considered, but are out of scope of this PR (for my own personal note-taking):

- Plutonium
- various offline conversion scripts (D&D Beyond; PDF conversion; misc. tooling; etc.)

TheGiddyLimit commented on node/generate-imperial-tags.js:
```diff
@@ -0,0 +1,42 @@
+// todo parse inches and weights
+const generateImperialForJson = (path) => {
+	console.log("\tGenerating imperial tags", path);
+	const fileContent = JSON.parse(fs.readFileSync(path, "utf8").replace(/^\uFEFF/, ""));
```
there's a `readJsonSync` somewhere

neupatryk replied:
Used

TheGiddyLimit commented on node/prettify-data/util-prettify-data.js:
```diff
@@ -4,7 +4,7 @@ import "../../js/parser.js";
 import "../../js/utils.js";
 import "../../js/utils-proporder.js";
 
-const FILE_BLOCKLIST = new Set([
```
should be moved to a shared constants file, not randomly exported

neupatryk replied:
Removed export

TheGiddyLimit commented on js/parser.js:
```diff
@@ -252,7 +252,7 @@ Parser.getAbilityModifier = function (abilityScore) {
 };
 
-Parser.getSpeedString = (ent, {isMetric = false, isSkipZeroWalk = false} = {}) => {
+Parser.getSpeedString = (ent, {isMetric = VetoolsConfig.get("styleSwitcher", "isMetric"), isSkipZeroWalk = false} = {}) => {
```
-> `isMetric = null` and 

```
isMetric ??= VetoolsConfig.get("styleSwitcher", "isMetric")
```

in the method body; avoid relying on `VetoolsConfig` being loaded when this method is defined

neupatryk replied:
Changed

TheGiddyLimit commented on node/generate-imperial-tags.js:
```diff
@@ -0,0 +1,42 @@
+
+	const prepareImperialTags = (value) => {
+		const cleanTags = /{@imperial\s(\S+\s[^}]+)}/gim;
+		const matchGroups = /(?<=\s|^)((?:(?:(?:\d+,)*\d+)(?:-|\\|\/))?(?:(?:\d+,)*\d+))\s(foot\b|feet\b|mile\b|miles\b|ft\.|mi\.)/gim;
```
use named capture groups `(?<unit>foot\b| ...`) -> `cleanedValue.replace(matchGroups, (...m) => `{@imperial ${m.last().value} ${m.last().unit}}`);`

neupatryk replied:
Using named groups

TheGiddyLimit commented on node/generate-imperial-tags.js:
```diff
@@ -0,0 +1,42 @@
+		return cleanedValue.replace(matchGroups, (_, value, unit) => `{@imperial ${value} ${unit}}`);
+	};
+
+	const parseValue = (value) => {
```
use `MiscUtil.getWalker`

neupatryk replied:
Used, great tool btw

TheGiddyLimit commented on node/generate-imperial-tags.js:
```diff
@@ -0,0 +1,42 @@
+	const preparedObject = Object.entries(fileContent).reduce(reduceObject, {});
+
+	fs.writeFileSync(path, JSON.stringify(preparedObject), "utf-8");
+	prettifyFile(path);
```
avoid prettifying in non-prettify scripts; use `CleanUtil.getCleanJson` when writing instead

neupatryk replied:
Using clean json

revilowaldow commented:
> `{@wm 60|ft}`

I think assuming ft where a unit is not otherwise provided would go a long way to reducing the extra text as well.

Personally much more in favour of the verbose flag than manually specifying single/plural which can just be ` if x == 1`

neupatryk commented:
> > `{@wm 60|ft}`
> 
> I think assuming ft where a unit is not otherwise provided would go a long way to reducing the extra text as well.
> 
> Personally much more in favour of the verbose flag than manually specifying single/plural which can just be ` if x == 1`

I would use verbose flag, but text in data is not consistent sometimes we have for example "60-foot" and sometimes "60-feet". I am trying to not change what is currently rendered in English text. If you think we don't need to maintain those edge cases, I can switch to verbose flag instead of singular|plural flag.

neupatryk commented:
> > `{@wm 60|ft}`
> 
> I think assuming ft where a unit is not otherwise provided would go a long way to reducing the extra text as well.
> 
> Personally much more in favour of the verbose flag than manually specifying single/plural which can just be ` if x == 1`

Also, you want to default to feet?

`{@wm 60}`  would be  `60 ft.`

And `{@wm 60|v}` would be `60 feet`
Or I missed the point?

neupatryk commented on js/render.js:
```diff
@@ -1785,6 +1785,54 @@ globalThis.Renderer = function () {
+				const isShortForm = !isSingularFlag && !isPluralFlag;
+
+				let isPlural;
+				const parsedValue = value.replace(/,/g, "").replace(/(\d+)(?:(\D)(\d+))?/gm, (_, v1, valueDelimiter, v2) => {
```
@TheGiddyLimit I am not sure about this part. You said that we don't want to regex in the render, but there is case where we have for example `60/120 feet` or `60-120 feet`. We need to parse both values to metric, but I don't have a good idea how we could store the split part without greatly increasing the complexity of tag. With this replace, we can split to number|delimiter|number and construct it again. Do you have any suggestions?

neupatryk commented on js/render.js:
```diff
@@ -1785,6 +1785,54 @@ globalThis.Renderer = function () {
 				textStack[0] += isNaN(amount) ? unitSingle : Number(amount) > 1 ? (unitPlural || unitSingle.toPlural()) : unitSingle;
 				break;
 			}
+			case "@wm": {
```
If you think code is in the wrong place, I would be happy to know how to structure it in the proper place. This case is quite large in comparison to others.

TheGiddyLimit replied:
all good for now

neupatryk commented:
> fine enough as a concept, though plenty to be picking over implementation-wise
> 
> > In my monster manual, conversion is 1mile = 1.5km.
> 
> what version is that, for reference? Happy to adjust 5etools' numbers, they're mostly extrapolated from the one Reddit thread [noted as a comment in the source](https://old.reddit.com/r/dndnext/comments/6gkuec). Screenshots/pictures/page numbers/ISBN(!)/... much appreciated
> 
> not a big fan of `@imperial`; given `@unit` is taken, I reckon `@wm` ("weights and measures"). Since these are going to end up absolutely everywhere, the shorter the better
> 
> not keen on your current tag format, though. I'd like to see something more structured, e.g.:
> 
> ```
> {@imperial 60 ft.} -> {@wm 60|ft} -> rendered as "60 ft."
> {@imperial 60 feet} -> {@wm 60|ft|p} // "p" = "plural" -> rendered as "60 feet"
> {@imperial 1 foot} -> {@wm 1|ft|s} // "s" = "singular" -> rendered as "1 foot"
> ```
> 
> or
> 
> ```
> {@imperial 60 ft.} -> {@wm 60|ft} -> rendered as "60 ft."
> {@imperial 60 feet} -> {@wm 60|ft|v} // "v" = "verbose"; explicitly converted to plural as `60` -> rendered as "60 feet"
> {@imperial 1 foot} -> {@wm 1|ft|v} // "v" = "verbose"; explicitly converted to singular as `1` -> rendered as "1 foot"
> ```
> 
> there should never be regex in the renderer; we expect to be rendering well-formed data, not parsing it
> 
> the above might have some holes, but the gist should be: splitting on `|` should yield:
> 
> * a single number
> * a single unit specifier, picked from a minimal set (i.e. we should **not** allow both `ft` and `feet`; stick to one specifier for each unit, and render it as required)
> * anything else we need (e.g. the rendering flags above; see `@area` for inspiration/prior work)
> 
> additional things which need considered:
> 
> * There was some recent chat about converting `monster.senses` to something more structured, which would eliminate the need for adding tags there. Would therefore prefer if those were _not_ tagged
> * Homebrew builder
> * Text converter
> 
> additional things which need considered, but are out of scope of this PR (for my own personal note-taking):
> 
> * Plutonium
> * various offline conversion scripts (D&D Beyond; PDF conversion; misc. tooling; etc.)

Changed to @wm tag
Split data to pipes
Monster senses are ignored in the walker

neupatryk commented:
> > In my monster manual, conversion is 1mile = 1.5km.
> 
> what version is that, for reference? Happy to adjust 5etools' numbers, they're mostly extrapolated from the one Reddit thread [noted as a comment in the source](https://old.reddit.com/r/dndnext/comments/6gkuec). Screenshots/pictures/page numbers/ISBN(!)/... much appreciated

I have monster manual from March 2020 with errata from 2018 included. Here are some pictures. ISBN 978-83-949332-3-4

Banshee detect life (p 21): English 5 miles, Polish 7.5 km
![448985719_386460364058934_3553745746367719254_n](https://github.com/user-attachments/assets/5959c7cd-ec4b-4efb-b8f0-f711b85baf28)

Aboleth enslave (p 13): English 1 mile, Polish 1.5km
![451847223_460237240136683_1925632951306656200_n](https://github.com/user-attachments/assets/025514a4-c40a-4f4d-a559-e0e6a682762d)

Reading this Reddit post, I don't see anything about miles conversion. Am I missing something?

neupatryk commented:
Some items have measurement in the name `Pole (10-foot)`. And it is breaking parsing for them. I disabled parsing name field, but looks like it is needed in that case.

neupatryk commented:
Current converting of kilometers will sometimes yield weird numbers 

![image](https://github.com/user-attachments/assets/6e88ed1e-4d2e-4f91-a2d4-db34b14e3f20)

TheGiddyLimit commented:
> {@wm 60} would be 60 ft.
> And {@wm 60|v} would be 60 feet
> Or I missed the point?

👍 that's perfect

---

>  text in data is not consistent sometimes we have for example "60-foot" and sometimes "60-feet"

I see. Then, can you dump a list of all the different forms you've found? We can hash out how the tag should work from there

---

> Reading this Reddit post, I don't see anything about miles conversion. Am I missing something?

nope, that's what I mean by "extrapolated." If there wasn't something in that Reddit post, I had to make something up. Pics of the values used in print are perfect, ty

--- 

> Some items have measurement in the name

yeah, leave those. Changing them would break `{@item ...}` links etc, which isn't worth the pain

---

> wonky numbers

yeah, that's a JavaScript Feature:tm:. Use e.g. `myNumberVar.toFixed(3)` to get a sensible string

revilowaldow commented:
> {@wm 60} would be 60 ft.
> And {@wm 60|v} would be 60 feet
> Or I missed the point?
> 👍 that's perfect

I was thinking more:

`{@wm 60}` 60 ft.
`{@wm 60|mi}` 60 mi.
`{@wm 60|mi|v}` 60 miles
`{@wm 60||v}` 60 feet

Basically:

`{@wm <quantity>|<unit>|<flags>|<optional custom text>}`

- Quantity is your original value
   - Gives you the number, but also singular/plural
- Unit tells you your original unit which informs:
   - what unit text to display
   - what conversion to use when showing metric (ft implies m, mi implies km, F implies C, lb implies kg etc)
   - when absent like a default source assume ft
- Flags combined similarly to ingredient or area flags
   - v Verbose, controls unit abbreviation 
   - w Word, number to word (one ft. rather than 1 ft.) 
   - c,f,r already exist for ceil floor round etc
   - s Singular, force singular display eg 60-foot vs 60-feet
   - d Dash, 60-ft. Vs 60 ft. 
   - x eXact, ignore rules simplification and provide real conversions to some decimal precision
   - etc, these are just thrown out there
   - Which way round defaults are for eg dashes can be sorted on the basis of what is found in the text
- Optional text, unsure quite how you'd use this, maybe a custom display if the flags aren't quite doing it for you, could even be followed up with `|custom metric text}`

TheGiddyLimit commented:
mb, missed that wonky `|v`

neupatryk commented:
@TheGiddyLimit 
Moved code to convert utils
Spell ranges are now converted

I was looking in to homebrew builder, but I have a hard time finding how indexed actions are stored. They don't have tags, even if I parse whole data.

neupatryk commented:
Also, these are some combinations that I saw in the data 

Number: 1 5 30 120 3000 1,200 300/1,200 50-90
Combinations examples 10-feet 10-foot 1,200-foot 200/800 ft. 10-ft.-by-10-ft. 
Tricky combinations 120 ft 5ft. 5feet 90\u2011foot
 
 I think tricky combinations are typos and should not be handled.

neupatryk commented:
@TheGiddyLimit @revilowaldow Is there some way to ignore item names that are linking to existing items? Look at this example
Carpet of flying have field items with some names. Those names are not tags, so texts are converted.
![image](https://github.com/user-attachments/assets/5bfa842d-0289-425b-a3dd-7255165d1c99)

When rendering item, they are converted to links and we have error.
![image](https://github.com/user-attachments/assets/cecf3d6c-3c98-42b1-aa97-e890b2ee4c8d)

revilowaldow commented:
Those are `item`s inside an `itemGroup` you should probably just ignore `itemGroup.[].items` entirely.

neupatryk commented:
> Those are `item`s inside an `itemGroup` you should probably just ignore `itemGroup.[].items` entirely.

@revilowaldow I know how to exclude key from the walker, but don't know how to do it with the path. Do you have some example?

TheGiddyLimit commented:
the walker _does_ give you the previous key (i.e. you can make a handler which is `string: (str, key) => { ... }`, but that's not necessarily accurate, since there could be any number of things called `items`

It does, however, have `pre`- and `post`-steps for each type, so you can track the keys yourself

Something like this should do it:

```js
(() => {
	const _EXAMPLE_OBJECT = {
		"itemGroup": [
			{
				"name": "Log Me",
				"items": [
					"Do Not Log Me"
				]
			}
		]
	}

	const walker = MiscUtil.getWalker();

	const keyStack = [];
	return walker.walk(
		_EXAMPLE_OBJECT,
		{
			preObject: (obj, key) => keyStack.push(key),
			postObject: () => keyStack.pop(),
			preArray: (arr, key) => keyStack.push(key),
			posArray: () => keyStack.pop(),

			string: (str) => {
				if (keyStack.at(-2) === "itemGroup" && keyStack.at(-1) === "items") return str;

				console.log(str)

				return str;
			},
		},
	);
})()
```

TheGiddyLimit commented:
> Tricky combinations 120 ft 5ft. 5feet 90\u2011foot
>
> I think tricky combinations are typos and should not be handled.

if you want to post a second PR fixing those typos, that can certainly be merged ahead of this one

TheGiddyLimit commented:
> I was looking in to homebrew builder, but I have a hard time finding how indexed actions are stored. They don't have tags, even if I parse whole data.

not sure I follow

either way, each homebrew converter/builder has a "post-processing" step somewhere (unfortunately not well factored-out, so you might have to do a bit of scrolling) where you can see other similar `...Tag` classes in use: 

- https://github.com/5etools-mirror-2/5etools-mirror-2.github.io/blob/main/js/converter-creature.js#L1426
- https://github.com/5etools-mirror-2/5etools-mirror-2.github.io/blob/main/js/makebrew-creature.js#L425

your new `WeightMeasureTag` should just be called in this step for each converter/builder

neupatryk commented:
> the walker _does_ give you the previous key (i.e. you can make a handler which is `string: (str, key) => { ... }`, but that's not necessarily accurate, since there could be any number of things called `items`
> 
> It does, however, have `pre`- and `post`-steps for each type, so you can track the keys yourself
> 
> Something like this should do it (example should execute in the console as-is):

Based on my tests, that solution is not working for any case when preObject is triggering on the top level key in the object. In items.json itemGroup is top key so preObject is not producing correct key because it relies on the last visited key not the current one.

This is result on the very simplified items.json 

![image](https://github.com/user-attachments/assets/5fa18a48-b8bb-473f-9b81-b86316ea0eeb)
![image](https://github.com/user-attachments/assets/cce6038f-0ff8-413a-bc93-c902cebf3e99)
![image](https://github.com/user-attachments/assets/54b7ff35-4564-49c6-b73d-7583ffebfdf0)

neupatryk commented:
> not sure I follow

When I am talking about indexed action, I mean about this.

![image](https://github.com/user-attachments/assets/482bba6f-ab4d-48a7-ac58-ad6afc39bfe8)

We have a list of actions that are not storing info about wm tag even when I parse whole data folder. I need to check if this is an issue of post-processing that you mentioned, but maybe you will know faster.

TheGiddyLimit commented on js/converterutils-entries.js:
```diff
@@ -390,6 +391,90 @@ class HazardTag {
+	}
+}
+// match 1 5 30 120 3000 1,200 300/1,200 50-90
+WeightMeasureTag._RE_NUMBER_SEE = "(?<value>(((\\d+,)*\\d+)(-|/|\\s(and|or|to)\\s))?((\\d+,)*\\d+))";
+// match " " or -
+WeightMeasureTag._RE_DELIMITER_SEE = "(?<delimiter>[\\s-])";
+// match unit only of it is whole word
+WeightMeasureTag._RE_UNIT_SEE = "(?<unit>ft\\.|mi\\.|lb(?=[\\.s])|foot\\b|feet\\b|mile[s]?\\b|pound[s]?\\b)";
```
prefer using `static _...` and pulling these into the class def (like you have `static _WALKER` above)

(there's a bunch of this in old code, because `static` didn't work reliably in old browsers/node, so don't worry about those)

neupatryk replied:
Done

TheGiddyLimit commented on js/filter-bestiary.js:
```diff
@@ -96,7 +96,12 @@ class PageFilterBestiary extends PageFilterBase {
 			itemSortFn: null,
 		});
-		this._speedFilter = new RangeFilter({header: "Speed", min: 30, max: 30, suffix: " ft"});
+		if (VetoolsConfig.get("styleSwitcher", "isMetric")) {
```
this needs reworked -- the underlying data will still be in feet, so you'll need to keep the old `_speedFilter` and add a `displayFn` to it; have the `displayFn` render the value in `m` if the user has the setting enabled

changing the underlying values of the filters has Bad:tm: effects, like making links via `@filter ...|speed=30` tags not work

neupatryk replied:
Done

TheGiddyLimit commented on js/filter-bestiary.js:
```diff
@@ -369,15 +374,18 @@ class PageFilterBestiary extends PageFilterBase {
 		if (typeof mon.speed === "number" && mon.speed > 0) {
 			mon._fSpeedType = ["walk"];
-			mon._fSpeed = mon.speed;
+			mon._fSpeed = isMetric ? Parser.metric.getMetricNumber({ originalValue: mon.speed, originalUnit: "ft" }) : mon.speed;
```
as above, needs to be kept in feet internally

TheGiddyLimit commented on js/filter-bestiary.js:
```diff
@@ -369,15 +374,18 @@ class PageFilterBestiary extends PageFilterBase {
 		mon._fSpeedType = Object.keys(mon.speed).filter(k => mon.speed[k]);
-		if (mon._fSpeedType.length) mon._fSpeed = mon._fSpeedType.map(k => mon.speed[k].number || mon.speed[k]).filter(it => !isNaN(it)).sort((a, b) => SortUtil.ascSort(b, a))[0];
-		else mon._fSpeed = 0;
+		if (mon._fSpeedType.length) {
```
as above

TheGiddyLimit commented on js/filter-items.js:
```diff
@@ -72,7 +72,12 @@ class PageFilterEquipment extends PageFilterBase {
 			labelDisplayFn: it => !it ? "None" : Parser.getDisplayCurrency(CurrencyUtil.doSimplifyCoins({cp: it})),
 		});
-		this._weightFilter = new RangeFilter({header: "Weight", min: 0, max: 100, isAllowGreater: true, suffix: " lb."});
+
```
same stuff as for the bestiary speed filter, etc.

neupatryk replied:
Done

TheGiddyLimit commented:
> I found it, in the top function first layer of keys is stripped for some reason. Do you have some advice on going forward with this? I could just run this tag in non stripped section, but I expect it is there for some reason.

hm, right enough. Given that this thing is going to have its own set of blocklisted paths, it should probably be run on its own. So instead of calling it inside `TagJson`, just call your `WeightMeasureTag.tryRun` right after where `TagJson` gets used; that should do just fine

TheGiddyLimit commented:
> We have a list of actions that are not storing info about wm tag even when I parse whole data folder. I need to check if this is an issue of post-processing that you mentioned, but maybe you will know faster.

I would expect it to Just Work:tm: with only the tagging step in post-processing, as:

- the user selects an action to add
- that action is added to the statblock
- the post-processing runs, and updates the statblock, converting distances to `@wm` where appropriate

so if it's _not_ doing that rn, then it's probably just because of the wonky tagging discussed above

neupatryk commented:
> hm, right enough. Given that this thing is going to have its own set of blocklisted paths, it should probably be run on its own. So instead of calling it inside `TagJson`, just call your `WeightMeasureTag.tryRun` right after where `TagJson` gets used; that should do just fine

I moved to parsing to lower section and it works now

neupatryk commented:
> I would expect it to Just Work:tm: with only the tagging step in post-processing, as:
> 
> * the user selects an action to add
> * that action is added to the statblock
> * the post-processing runs, and updates the statblock, converting distances to `@wm` where appropriate
> 
> so if it's _not_ doing that rn, then it's probably just because of the wonky tagging discussed above

I need more testing, but when I commented all tag parsing in the mentioned files, standard actions are still inserted with tags.
